File: 213-house-robber-ii/213-house-robber-ii.kt
class Solution {
    
    val mem = mutableMapOf<String, Int>()
    
    fun rob(nums: IntArray): Int {
        if(nums.size==1){
            return nums[0]
        }
        var maxRobberyAmount = 0
        for(i in 0..nums.size){
            maxRobberyAmount = maxOf(maxRobberyAmount, maxRob(nums, i, i))
        }
        return maxRobberyAmount
    }
    
    fun maxRob(nums: IntArray, current:Int, start:Int):Int{
        var currentPos = current.toString() + "-" + start.toString()
        if(mem[currentPos]!=null){
            return mem[currentPos] as Int
        }
        
        if(current>nums.size-1){
            return 0
        }
        if(current==nums.size-1){
            if(start==0){
                return 0
            }
        }
            mem[currentPos] = nums[current]+maxOf(maxRob(nums, current+2, start), maxRob(nums, current+3, start))
            return mem[currentPos] as Int
        
        
    }
}

File: 213-house-robber-ii/213-house-robber-ii.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        
        if len(nums)==1:
            return nums[0]
        
        dp = {}
        
        def maxRob(start, end):
            
            if (start, end) in dp:
                return dp[(start, end)]
            
            if end>len(nums)-1:
                return 0
            
            if end==len(nums)-1:
                if start==0:
                    return 0
            
            dp[(start, end)] = nums[end] + max(maxRob(start, end+2), maxRob(start, end+3))
            return dp[(start, end)]
        
        maxAmount = 0
        for i in range(len(nums)):
            maxAmount = max(maxAmount, maxRob(i,i))
        
        return maxAmount
        